sig-node/openharmony/implementation/KubeEdge Runs On OpenHarmony.md

```diff
@@ -0,0 +1,1277 @@
+# Ensure the OpenHarmony is networked then execute hello-world
+docker run hello-world
+
+# You can see
+Unable to find image 'hello-world:latest' locally
+latest: Pulling from library/hello-world
+93288797bd35: Pull complete 
+Digest: sha256:2498fce14358aa50ead0cc6c19990fc6ff866ce72aeb5546e1d59caac3d0d60f
+Status: Downloaded newer image for hello-world:latest
+
+Hello from Docker!
+This message shows that your installation appears to be working correctly.
+
+To generate this message, Docker took the following steps:
+ 1. The Docker client contacted the Docker daemon.
+ 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
+    (arm64v8)
+ 3. The Docker daemon created a new container from that image which runs the
+    executable that produces the output you are currently reading.
+ 4. The Docker daemon streamed that output to the Docker client, which sent it
+    to your terminal.
+
+To try something more ambitious, you can run an Ubuntu container with:
+ $ docker run -it ubuntu bash
+
+Share images, automate workflows, and more with a free Docker ID:
+ https://hub.docker.com/
+
+For more examples and ideas, visit:
+ https://docs.docker.com/get-started/
+```
+
+
+
+## Key steps for OpenHarmony to run KubeEdge
+
+
+
+Use KubeEdge to bridge the cloud and the OpenHarmony edge to achieve cloud-side collaboration.
+
+
+
+##### 1. Install KubeEdge cloudcore on the cloud and get token
+
+##### 2. KubeEdge edgecore source static compilation
+
+Compile KubeEdge source code on Arm server to obtain edgecore static binary file.
+
+```
+# 1. Match OpenHamrony's memory.stat format,
+
+# This is actually a third-party library that fails to include all possible memory.stat formats, which I have submitted to issues to fix. https://github.com/opencontainers/runc/issues/3643
+
+# You can also modify /vendor/github.com/opencontainers/runc/libcontainer/cgroups/fscommon/utils.go
+
+func ParseKeyValue(t string) (string, uint64, error) {
+
+	tmp := strings.Replace(t, ":", "", -1)
+	tmpo := strings.Replace(tmp, "\t", "", -1)
+	tmpt := strings.Replace(tmpo, " kB", "", -1)
+	count := strings.Count(tmpt, " ")
+	tmps := strings.Replace(tmpt, " ", "", count-1)
+
+	parts := strings.SplitN(tmps, " ", 3)
+	if len(parts) != 2 {
+		return "", 0, fmt.Errorf("line %q is not in key value format", t)
+	}
+
+	value, err := ParseUint(parts[1], 10, 64)
+	if err != nil {
+		return "", 0, fmt.Errorf("unable to convert to uint64: %v", err)
+	}
+
+	return parts[0], value, nil
+}
+
+
+# 2. Since openharmony uses overlay2 to install docker, you need to modify the path to the kubelet file in kubeedge to make it the same as the path to docker overlay2.
+
+# 3. Compilation edgecore
+
+cd kubeedge-1.9.1
+docker build -t kubeedge/edgecore:v1.9.1 -f build/edge/Dockerfile .
+docker cp $(docker create --rm kubeedge/edgecore:v1.9.1):/usr/local/bin/edgecore ./edgecore.1.9.1
+
+# 4. In the kubeedge directory there is edgecore.tag executable file copied to the openharmony board /bin/ 
```

Comment:
suggest: Copy the edgecore.tag executable file in the KubeEdge directory to the openharmony board `/bin/`.